0010-timestamps.adoc

```diff
@@ -0,0 +1,78 @@
+= Event timestamps =
+Tobin C. Harding <tobin.harding@coblox.tech>;
+:toc:
+:revdate: 2019-07-24
+
+NOTE: Author: {authors} +
+Date: {revdate} +
+Tracking issue: https://github.com/comit-network/comit-rs/issues/1044
+
+== Context
+
+Currently events that occur during COMIT protocol interactions are not timestamped.
+It would be nice to have time information added to events as they occur - for this we can use timestamps.
+
+== Assumptions
+
+It is assumed that time is from the perspective of the cnd i.e., events are ordered as they appear to have occurred to the cnd.
+This implies that the ordering of events is only of use to agents that interface with cnd via the HTTP API and possibly to the cnd.
+
+== Definition
+
+A timestamp is a sequence of characters or encoded information identifying when a certain event occurred ^1^.
+Timestamps are described by the ISO 8601 standard ^2^.
+
+* ^1^ https://en.wikipedia.org/wiki/Timestamp
+* ^2^ https://en.wikipedia.org/wiki/ISO_8601
+
+== Format
+
+Any representation of time will suffice.
+Considerations are:
+
+ - granularity
+ - representation i.e., wall clock time, seconds since epoch.
+
+== Events to timestamp
+
+Events that could potentially be timestamped.
+
+=== Events from Alice's perspective
+
+==== happy swap
+- swap accept received (fund available)
+- Alpha ledger funded (refund available)
+- Beta ledger funded (redeem available)
+- Beta asset redeemed (by herself)
+- Alpha asset redeemed (by Bob)
+
+==== swap declined
+- swap request sent
+- swap decline received
+
+=== Events from Bob's perspective
+
+==== happy swap
+- swap request received
+- swap accept sent
+- Alpha ledger funded (fund available)
+- Beta ledger funded (refund available)
+- Beta asset redeemed (by Alice, redeem available)
+- Alpha asset redeemed (by himself)
+
+====  decline swap
+- swap request received
+- swap decline sent
+
+== Suggestions
+
+I suggest we timestamp the events above using seconds since epoch.
```

Comment:
So if I understand correctly, your suggestion is to only timestamp the event from the moment `cnd` is made aware of them and ignore any timestamp coming from the blockchain.

It sounds reasonable but did you consider using timestamps from the blockchain?
E.g, when a `fund` transaction is seed, instead of saving the timestamp of when cnd is made aware, use the timestamp of the blockchain transaction (which is actually the timestamp of the block in which the transaction was confirmed).

Also, can we think about what would happen if we were to consider 6 confirmations for the `fund` event to be considered done?

We would have to assume that btsieve would return the transaction once all confirmations are done. So 1 hour after the transaction is broadcasted. Is that fine?

Reply:
Question: What do we want to achieve with the timestamp?
If the sole purpose is, to put the event in order (so that comit-i can display them in the correct sequence) I think a timestamp by the comit-node would be sufficient. 

If we want to depict block confirmations in the timestamp then this would actually be a different timestamp for me. 
Do we have other use-cases besides the event order in comit-i that involve the timestamp?

Reply:
I limited the scope of these timestamps to just being of use to clients i.e. comit-i (stated in the document).  I do not think we should bother with blockchain times.

Reply:
> I limited the scope of these timestamps to just being of use to clients i.e. comit-i (stated in the document). I do not think we should bother with blockchain times.

Can you add that to the spike then please?

Reply:
Added 'I suggest that all timestamps are created by the cnd since they are only useful to the cnd I feel it adds no real benefit to get verified times from the blockchain.'

Reply:
Did patch with a this added as a 'suggestion'.

Reply:
Getting timestamp from the blockchain is not all about getting **verified** times, it's also about getting all events especially if we implement the confirmation feature.

Please note that I agree with your proposal, I just want to not leave any stones unturned.